### PR TITLE
*: lock the temp index key properly in DML

### LINKED ISSUE #41515: Pessimistic lock not found when adding index
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

It happens occasionally on `realtikvtest/addindextest` of `check_dev_2`.

https://do.pingcap.net/jenkins/blue/organizations/jenkins/pingcap%2Ftidb%2Fghpr_check2/detail/ghpr_check2/14284/pipeline/231/

### 2. What did you expect to see? (Required)

No error.

### 3. What did you see instead (Required)

```
[2023/02/16 00:56:54.698 +08:00] [WARN] [session.go:1019] ["commit failed"] ["finished txn"="Txn{state=invalid}"] [error="tikv aborts txn: Error(Txn(Error(Mvcc(Error(PessimisticLockNotFound { start_ts: TimeStamp(439479229396549638), key: [116, 128, 0, 0, 0, 0, 0, 0, 162, 95, 105, 255, 255, 0, 0, 0, 0, 0, 1, 4, 0, 0, 0, 0, 0, 0, 0, 3] })))))"] [errorVerbose="tikv aborts txn: Error(Txn(Error(Mvcc(Error(PessimisticLockNotFound { start_ts: TimeStamp(439479229396549638), key: [116, 128, 0, 0, 0, 0, 0, 0, 162, 95, 105, 255, 255, 0, 0, 0, 0, 0, 1, 4, 0, 0, 0, 0, 0, 0, 0, 3] })))))\ngithub.com/tikv/client-go/v2/error.ExtractKeyErr\n\texternal/com_github_tikv_client_go_v2/error/error.go:301\ngithub.com/tikv/client-go/v2/txnkv/txnlock.ExtractLockFromKeyErr\n\texternal/com_github_tikv_client_go_v2/txnkv/txnlock/lock.go:27\ngithub.com/tikv/client-go/v2/txnkv/transaction.actionPrewrite.handleSingleBatch\n\texternal/com_github_tikv_client_go_v2/txnkv/transaction/prewrite.go:387\ngithub.com/tikv/client-go/v2/txnkv/transaction.(*twoPhaseCommitter).doActionOnBatches\n\texternal/com_github_tikv_client_go_v2/txnkv/transaction/2pc.go:1037\ngithub.com/tikv/client-go/v2/txnkv/transaction.(*twoPhaseCommitter).doActionOnGroupMutations\n\texternal/com_github_tikv_client_go_v2/txnkv/transaction/2pc.go:1018\ngithub.com/tikv/client-go/v2/txnkv/transaction.(*twoPhaseCommitter).doActionOnMutations\n\texternal/com_github_tikv_client_go_v2/txnkv/transaction/2pc.go:788\ngithub.com/tikv/client-go/v2/txnkv/transaction.(*twoPhaseCommitter).prewriteMutations\n\texternal/com_github_tikv_client_go_v2/txnkv/transaction/prewrite.go:439\ngithub.com/tikv/client-go/v2/txnkv/transaction.(*twoPhaseCommitter).execute\n\texternal/com_github_tikv_client_go_v2/txnkv/transaction/2pc.go:1482\ngithub.com/tikv/client-go/v2/txnkv/transaction.(*KVTxn).Commit\n\texternal/com_github_tikv_client_go_v2/txnkv/transaction/txn.go:496\ngithub.com/pingcap/tidb/store/driver/txn.(*tikvTxn).Commit\n\tstore/driver/txn/txn_driver.go:95\ngithub.com/pingcap/tidb/session.(*LazyTxn).Commit\n\tsession/txn.go:425\ngithub.com/pingcap/tidb/session.(*session).commitTxnWithTemporaryData\n\tsession/session.go:843\ngithub.com/pingcap/tidb/session.(*session).doCommit\n\tsession/session.go:729\ngithub.com/pingcap/tidb/session.(*session).retry\n\tsession/session.go:1300\ngithub.com/pingcap/tidb/session.(*session).doCommitWithRetry\n\tsession/session.go:1001\ngithub.com/pingcap/tidb/session.(*session).CommitTxn\n\tsession/session.go:1104\ngithub.com/pingcap/tidb/session.autoCommitAfterStmt\n\tsession/tidb.go:295\ngithub.com/pingcap/tidb/session.finishStmt\n\tsession/tidb.go:257\ngithub.com/pingcap/tidb/session.runStmt\n\tsession/session.go:2380\ngithub.com/pingcap/tidb/session.(*session).ExecuteStmt\n\tsession/session.go:2218\ngithub.com/pingcap/tidb/testkit.(*TestKit).ExecWithContext\n\ttestkit/testkit.go:323\ngithub.com/pingcap/tidb/testkit.(*TestKit).Exec\n\ttestkit/testkit.go:296\ngithub.com/pingcap/tidb/tests/realtikvtest/addindextest.updateWorker\n\ttests/realtikvtest/addindextest/workload.go:211\ngithub.com/pingcap/tidb/tests/realtikvtest/addindextest.(*worker).run\n\ttests/realtikvtest/addindextest/workload.go:104\nruntime.goexit\n\tGOROOT/src/runtime/asm_amd64.s:1594\ngithub.com/pingcap/errors.AddStack\n\texternal/com_github_pingcap_errors/errors.go:174\ngithub.com/pingcap/errors.Trace\n\texternal/com_github_pingcap_errors/juju_adaptor.go:15\ngithub.com/pingcap/tidb/store/driver/error.ToTiDBErr\n\tstore/driver/error/error.go:166\ngithub.com/pingcap/tidb/store/driver/txn.extractKeyErr\n\tstore/driver/txn/error.go:162\ngithub.com/pingcap/tidb/store/driver/txn.(*tikvTxn).extractKeyErr\n\tstore/driver/txn/txn_driver.go:310\ngithub.com/pingcap/tidb/store/driver/txn.(*tikvTxn).Commit\n\tstore/driver/txn/txn_driver.go:96\ngithub.com/pingcap/tidb/session.(*LazyTxn).Commit\n\tsession/txn.go:425\ngithub.com/pingcap/tidb/session.(*session).commitTxnWithTemporaryData\n\tsession/session.go:843\ngithub.com/pingcap/tidb/session.(*session).doCommit\n\tsession/session.go:729\ngithub.com/pingcap/tidb/session.(*session).retry\n\tsession/session.go:1300\ngithub.com/pingcap/tidb/session.(*session).doCommitWithRetry\n\tsession/session.go:1001\ngithub.com/pingcap/tidb/session.(*session).CommitTxn\n\tsession/session.go:1104\ngithub.com/pingcap/tidb/session.autoCommitAfterStmt\n\tsession/tidb.go:295\ngithub.com/pingcap/tidb/session.finishStmt\n\tsession/tidb.go:257\ngithub.com/pingcap/tidb/session.runStmt\n\tsession/session.go:2380\ngithub.com/pingcap/tidb/session.(*session).ExecuteStmt\n\tsession/session.go:2218\ngithub.com/pingcap/tidb/testkit.(*TestKit).ExecWithContext\n\ttestkit/testkit.go:323\ngithub.com/pingcap/tidb/testkit.(*TestKit).Exec\n\ttestkit/testkit.go:296\ngithub.com/pingcap/tidb/tests/realtikvtest/addindextest.updateWorker\n\ttests/realtikvtest/addindextest/workload.go:211\ngithub.com/pingcap/tidb/tests/realtikvtest/addindextest.(*worker).run\n\ttests/realtikvtest/addindextest/workload.go:104\nruntime.goexit\n\tGOROOT/src/runtime/asm_amd64.s:1594"]
```

### 4. What is your TiDB version? (Required)

d0919b907a

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close #41515

Problem Summary:

### What is changed and how it works?

- Add the unique check for temp index in `keyNeedToLock()`
- Refine some of the function parameters

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: tablecodec/tablecodec.go [hunk 1]
```diff
@@ -1345,10 +1345,10 @@ func (v *TempIndexValueElem) DecodeOne(b []byte, isCommonHandle bool) (remain []
 	case TempIndexValueFlagDeleted:
 		hLen := (uint16(b[0]) << 8) + uint16(b[1])
 		b = b[2:]
-		if isCommonHandle {
-			v.Handle, _ = kv.NewCommonHandle(b[:hLen])
+		if hLen == 8 {
```
Comment (Benjamin2037) [inline tablecodec/tablecodec.go:1348]: Can this 8 change to one const or only one variable?

File: session/txn.go [hunk 2]
```diff
@@ -635,6 +635,16 @@ func keyNeedToLock(k, v []byte, flags kv.KeyFlags) bool {
...
 	}
 
+	if tablecodec.IsTempIndexKey(k) {
+		tmpVal, err := tablecodec.DecodeTempIndexValue(v)
+		if err != nil {
+			logutil.BgLogger().Warn("decode temp index value failed", zap.Error(err))
+			return false
+		}
+		current := tmpVal.Current()
+		return current.Handle != nil || tablecodec.IndexKVIsUnique(current.Value)
```
Comment (Benjamin2037) [inline session/txn.go:645]: what is this retrun condition means? is uk or ?

**Review by hawkingrei — APPROVED** [review]

**Review by Benjamin2037 — APPROVED** [review]
LGTM

File: tablecodec/tablecodec.go [hunk 3]
```diff
@@ -1058,6 +1058,9 @@ func IsUntouchedIndexKValue(k, v []byte) bool {
 		return false
 	}
 	vLen := len(v)
+	if IsTempIndexKey(k) {
+		return vLen > 0 && v[vLen-1] == kv.UnCommitIndexKVFlag
```
[the remark this answers is not in the document]
Comment (tangenta) [reply tablecodec/tablecodec.go:1062]: No, temp index values always end with 'b'/'m'/'d'/'1'.

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**